### PR TITLE
Support for new versions of XEP-0313 MAM

### PR DESCRIPTION
**Pull request by arcusfelis**
All changes:
- Support for MAM v0.3, v0.4.1, v0.5. There are no major changes in v0.5.
- A client should specify appropriate namespace for IQ's he sends. After that a server decides which version should be used.
- Override is_complete_packet/3 (for cases when you want to choose what to archive).
- Make `<archived/>` element configurable.
- Fix access rights for MAM MUC (anyone was able to read the archive).
- no-store hint (client can add it to disable storing message).
- mam_SUITE was extended to support three MAM versions.
- Prefs backends were updated to support not-implemented error.
- ejabberd.cfg config change: `{access, mam_lookup_messages, [{default, all}]}.` becomes optional. The old variant is supported too (i.e. no need for config update).
- filter_room_packet was extended with more arguments.

| namespace | XEP version | Result stanzas | Query using forms |
|-------------|---------------|-------------|-----|
| urn:xmpp:mam:tmp | 0.2 | messages, iq with rsm | no |
| urn:xmpp:mam:0 | 0.3 | iq, messages, fin message with rsm | yes |
| urn:xmpp:mam:1 | 0.4.1 and 0.5 | messages, fin iq with rsm | yes |

Spec diffs

- http://xmpp.org/extensions/diff/api/xep/0313/diff/0.2/vs/0.3
- http://xmpp.org/extensions/diff/api/xep/0313/diff/0.3/vs/0.4.1
- http://xmpp.org/extensions/diff/api/xep/0313/diff/0.4.1/vs/0.5

### CODE REVIEW

**Comment by arcusfelis** [comment]
TODO next. Archive this for mod_mam_muc based on XEP v0.5. We can patch packet adding this element before storing. After extracting we can remove it depending on access rights.

```erlang
        <x xmlns='http://jabber.org/protocol/muc#user'>
          <item affiliation='none'
                jid='witch2@shakespeare.lit'
                role='participant' />
```

**Comment by arcusfelis** [comment]
Cherry-picked https://github.com/esl/MongooseIM/pull/647 into this branch (because we reached 48 minutes limit)

**Comment by michalwski** [comment]
I see that `<archived>` is used only in version 0.2. If I understand @arcusfelis's changes correctly, this element will be added for other clients using newer version if we configure the server to do it. @arcusfelis , correct me if I'm wrong.

**Comment by arcusfelis** [comment]
By default it's enabled for all clients.
But it's possible to disable it using config for all clients.

**Comment by michalwski** [comment]
I think this should be at least disabled by default or determined in some other way wether to add this tag or not.

**Comment by arcusfelis** [comment]
Fine. We can disable it by default.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 1]
```diff
@@ -714,6 +742,39 @@ can_send_to_conference(From, StateData) ->
     orelse
     is_allowed_nonparticipant(From, StateData).
 
+can_read_conference(UserJID,
```
Comment (arcusfelis) [inline apps/ejabberd/src/mod_muc_room.erl]: Looks scary

**Comment by arcusfelis** [comment]
It works for Kaiwa (stanza.io 7.3.4). It uses MAM v0.3.

**Pull request merged**